bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/actions/AVMFritzHeatingActions.java

```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public void setThingHandler(@Nullable ThingHandler handler) {
+        this.handler = (AVMFritzHeatingActionsHandler) handler;
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return handler;
+    }
+
+    @RuleAction(label = "@text/setBoostModeModeActionLabel", description = "@text/setBoostModeActionDescription")
+    public void setBoostMode(
+            @ActionInput(name = "Duration", label = "@text/setBoostModeDurationInputLabel", description = "@text/setBoostModeDurationInputDescription", type = "java.lang.Long", required = true) Long duration) {
+        AVMFritzHeatingActionsHandler actionsHandler = handler;
+        if (actionsHandler == null) {
+            logger.warn("AVMFritzHeatingActions ThingHandler is null!");
+            return;
+        }
+        actionsHandler.setBoostMode(duration.longValue());
+    }
+
+    public static void setBoostMode(@Nullable ThingActions actions, Long duration) {
+        invokeMethodOf(actions).setBoostMode(duration);
+    }
+
+    @RuleAction(label = "@text/setWindowOpenModeActionLabel", description = "@text/setWindowOpenModeActionDescription")
+    public void setWindowOpenMode(
+            @ActionInput(name = "Duration", label = "@text/setWindowOpenModeDurationInputLabel", description = "@text/setWindowOpenModeDurationInputDescription", type = "java.lang.Long", required = true) Long duration) {
+        AVMFritzHeatingActionsHandler actionsHandler = handler;
+        if (actionsHandler == null) {
+            logger.warn("AVMFritzHeatingActions ThingHandler is null!");
+            return;
+        }
+        actionsHandler.setWindowOpenMode(duration.longValue());
+    }
+
+    public static void setWindowOpenMode(@Nullable ThingActions actions, Long duration) {
+        invokeMethodOf(actions).setWindowOpenMode(duration);
+    }
+
+    private static AVMFritzHeatingActions invokeMethodOf(@Nullable ThingActions actions) {
+        if (actions == null) {
+            throw new IllegalArgumentException("Actions cannot be null");
+        }
+        if (actions.getClass().getName().equals(AVMFritzHeatingActions.class.getName())) {
+            if (actions instanceof AVMFritzHeatingActions) {
+                return (AVMFritzHeatingActions) actions;
+            } else {
+                return (AVMFritzHeatingActions) Proxy.newProxyInstance(AVMFritzHeatingActions.class.getClassLoader(),
+                        new Class[] { AVMFritzHeatingActions.class }, (Object proxy, Method method, Object[] args) -> {
+                            Method m = actions.getClass().getDeclaredMethod(method.getName(),
+                                    method.getParameterTypes());
+                            return m.invoke(actions, args);
+                        });
+            }
+        }
+        throw new IllegalArgumentException("Actions is not an instance of AVMFritzHeatingActions");
+    }
```

Comment:
Proxy instances can only be interfaces, not concrete classes. So you need to make an interface that has all of your methods your actions are going to use.

This `invokeMethodOf` method should likewise return an interface instead.

Reply:
Isn't that the pattern that was just done on ALL our thing actions in this repo?

Reply:
> Isn't that the pattern that was just done on ALL our thing actions in this repo?

It is, but the pattern wasn't done correctly in this particular instance. All of the other ones (or at least the ones that I approved) require an additional interface class to be created.

Reply:
Done.